qwerty/ast_visitor.hpp

```diff
@@ -315,6 +315,60 @@ struct ClassicalTypeCheckVisitor : BaseTypeCheckVisitor {
     virtual bool visit(ASTVisitContext &ctx, Slice &slice) override;
 };
 
+// Rewrites AST w/o syntax sugar. While technically a form of canonicalization,
+// this visit pass is made distinct to ensure only a very small set of transformations
+// can avoid type-checking.
+struct DesugarVisitor : ASTVisitor {
```

Comment:
You can make this shorter by making it subclass `ObliviousASTVisitor` instead. See `FlagImmaterialVisitor` for an example of this.

This isn't functionally wrong though. And much of the existing visitors still have many, many `{ return true; }`s... I'm no saint

Reply:
I pushed up a slightly different way of doing it